refactor/move_frontend_signing into develop

dhelmert opened the pull request:
### Reason for this PR
- Fronted signing gets a new workflow, which needs to be incorporated into the api

### Changes in this PR
- Added frontend signing support for rejection of operations
- Moved submitting signed proposal and transaction completely to operation

dhelmert committed:
Refactor operation logic to move all submitSignedTransaction and getUnsignedTransaction to Operation
Merged getOperation methods into one, with more filters
Add operationId creation description

Co-authored-by: Niko Bergemann <niko.bergemann@web.de>
Co-authored-by: Anemone Kampkötter <anemone@kampkoetter.de>
Co-authored-by: Bastian Haverkamp <bastihav@mail.upb.de>
Co-authored-by: Luca Schäfers <lrschaef@mail.uni-paderborn.de>
Co-authored-by: Matthias Geuchen <mgeuchen@mail.uni-paderborn.de>
Co-authored-by: Marie-Christin Oczko <mco33415@mail.uni-paderborn.de>

NikoBergemann committed:
apply changes to "getOperations"filter

Co-authored-by: Daniel Helmert <dhelmert@mail.uni-paderborn.de>

NikoBergemann commented on hlf/chaincode/operation.md:
```diff
@@ -6,9 +6,10 @@ The Errors returned are defined [here](errors.md#Errors).
 
-### ApproveTransaction
+### ProposeTransaction
 - ID = approveTransaction
```
```suggestion
- ID = proposeTransaction

```

NikoBergemann commented on hlf/chaincode/operation.md:
```diff
@@ -119,19 +104,26 @@ The Errors returned are defined [here](errors.md#Errors).
 
 ### <a id="OperationData" />OperationData
+The operationId in OperationData is constructed as follows:  
+operationId = SHA256($contractName:$transactionName:$parameters)  
```
```suggestion
operationId = String(Base64(SHA256($contractName:$transactionName:$parameters))) 

```

matthias-geuchen committed:
Update access rights

Co-authored-by: Daniel Helmert <dhelmert@mail.uni-paderborn.de>
Co-authored-by: Robin Helmert <rhelmert@mail.uni-paderborn.de>
Co-authored-by: Niko Bergemann <niko.bergemann@web.de>
Co-authored-by: Marie-Christin Oczko <mco33415@mail.uni-paderborn.de>
Co-authored-by: Steffen Luis <sluis1@mail.uni-paderborn.de>
Co-authored-by: Anemone Kampkötter <anemone@kampkoetter.de>

dhelmert committed:
Update hlf/chaincode/operation.md

Co-authored-by: Niko Bergemann <nikob@mail.uni-paderborn.de>

dhelmert committed:
Update hlf/chaincode/operation.md

Co-authored-by: Niko Bergemann <nikob@mail.uni-paderborn.de>

dhelmert committed:
Rename ProposeOperation to initiateOperation

Co-authored-by: Niko Bergemann <niko.bergemann@web.de>
Co-authored-by: Anemone Kampkötter <anemone@kampkoetter.de>

dhelmert committed:
Update general communication, to reflect the new rejection logic

Co-authored-by: Niko Bergemann <niko.bergemann@web.de>
Co-authored-by: Anemone Kampkötter <anemone@kampkoetter.de>

dhelmert committed:
Change state filter to states
Split operation approval with implicit execution into two parts; approveOperation and executeTransaction

Co-authored-by: Robin Helmert <rhelmert@mail.uni-paderborn.de>
Co-authored-by: Niko Bergemann <niko.bergemann@web.de>
Co-authored-by: Marie-Christin Oczko <mco33415@mail.uni-paderborn.de>
Co-authored-by: Steffen Luis <sluis1@mail.uni-paderborn.de>
Co-authored-by: Anemone Kampkötter <anemone@kampkoetter.de>

NikoBergemann committed:
Update operation.md

NikoBergemann committed:
Update general-communication.md

add timeout to executeTransaction()

bastihav commented on hlf/chaincode/group.md:
```diff
@@ -10,6 +10,9 @@ The Errors returned are defined [here](errors.md#Errors).
 - ID = addUserToGroup
+- Required Approvals
+  - Users
+    - lagom-scala-admin-org1 (environment variable)
```
this is not an environment variable anymore, right?

NikoBergemann replied:
```suggestion
    - lagom-scala-admin-org1 (hyperledger attr. (set when enrolling))
```

bastihav commented on hlf/chaincode/operation.md:
```diff
@@ -75,22 +60,16 @@ The Errors returned are defined [here](errors.md#Errors).
-
-### GetOperationData
-- ID = getOperationData
+### RejectOperation
```
should throw an error if the user that wants to reject is not on the missing approval list/groups

bastihav replied:
see #106

bastihav commented on hlf/scala/general-communication.md:
```diff
@@ -65,35 +65,56 @@ The ADMIN can then
-(e.g. an InsufficientApproval Error, or information about the successfully changed ledger state)
+The approvalResult is a JsonObject [OperationData](../chaincode/operation.md#OperationData). 
+Invoking executeTransactionFromOperation executes the transaction described by the transactionInfo in the OperationData.
+In case of a NetworkException, the transaction is retried until the specified timeout is reached.
```
on implementation side, how is this implemented? Do you just retry the whole time? (the call might terminate in microseconds if the socket is not available, so this is really really bad busy waiting, is there a timeout between tries?)

NikoBergemann replied:
this is not implemented at all, and we have not even considered pros and cons of different approaches.
Regarding a timeout after X attempts - I was specifically told to not do that

bastihav commented on hlf/scala/general-communication.md:
```diff
@@ -65,35 +65,56 @@ The ADMIN can then
-An attempt has been made to execute the "real" transaction (e.g. "AddCertificate")
+If submitSignedTransaction succeeds, the approval of the User has been stored on the OperationContract.  
+If executeTransaction succeeds, the "real" transaction (e.g. "AddCertificate") is executed.  
+If the approval or the "real" transaction fails due to the transaction itself being invalid (e.g. due to changed ledger state), the operation is rejected.
```
rejected by? the system? the chain? what enrollmentId/grou?

NikoBergemann replied:
```suggestion
If the approval or the "real" transaction fails due to the transaction itself being invalid (e.g. due to changed ledger state), the operation is rejected by the chain.
```

bastihav commented on hlf/chaincode/operation.md:
```diff
@@ -119,19 +106,27 @@ The Errors returned are defined [here](errors.md#Errors).
 
 ### <a id="OperationData" />OperationData
+The operationId in OperationData is constructed as follows:  
+operationId = String(Base64(SHA256($contractName:$transactionName:$parameters))) // parameters is stripped of all spaces (replace(" ", ""))
```
```suggestion
operationId = String(Base64Url(SHA256($contractName:$transactionName:$parameters))) // parameters is stripped of all spaces (replace(" ", ""))

```

NikoBergemann committed:
Update hlf/chaincode/operation.md

Co-authored-by: Bastian Haverkamp <bastihav@mail.upb.de>

bastihav committed:
Update hlf/chaincode/group.md

Co-authored-by: Niko Bergemann <nikob@mail.uni-paderborn.de>

bastihav committed:
Update hlf/scala/general-communication.md

Co-authored-by: Niko Bergemann <nikob@mail.uni-paderborn.de>

NikoBergemann committed:
merge from develop

Co-authored-by: Matthias Geuchen <mgeuchen@mail.uni-paderborn.de>
Co-authored-by: Marie-Christin Oczko <mco33415@mail.uni-paderborn.de>
Co-authored-by: Steffen Luis <sluis1@mail.uni-paderborn.de>
Co-authored-by: Anemone Kampkötter <anemone@kampkoetter.de>
